src/features/smartSearch/components/filters/EmailHistory/index.tsx

```diff
@@ -177,11 +177,16 @@ const EmailHistory = ({
                 onChange={(e) => setValueToKey('operator', e.target.value)}
                 value={filter.config.operator}
               >
-                {Object.values(MESSAGE_KEY_BY_OP).map((operator) => (
-                  <MenuItem key={operator} value={operator}>
-                    <Msg id={localMessageIds.operatorSelect[operator]} />
-                  </MenuItem>
-                ))}
+                {Object.keys(MESSAGE_KEY_BY_OP).map((operator) => {
+                  const opKey = operator as keyof typeof MESSAGE_KEY_BY_OP;
```

Comment:
I'm just curious about why we had to casts each operator `as keyof typeof` MESSAGE_KEY_BY_OP? 👀

Reply:
Great question!

This is (IMO) because of a shortcoming in Typescript. The object is `MESSAGE_KEY_BY_OP` is declared `as const` (on lines 24-29), which means that it's not possible to add any keys to it after it is declared. This means that in most situations, typescript knows that the keys on this object can only be `not_opened`, `not_sent`, `opened` or `sent`.

But the type declarations for `Object.keys()` does not understand this. If it did understand it, the return type of `Object.keys()` would be `('not_opened', 'not_sent', 'opened', 'sent')[]`. I.e. an array of strings but only those four specific strings.

But the return type of `Object.keys()` is just `string[]`, i.e. an array of any strings. So on line 82 where I'm trying to use the key to actually access `MESSAGE_KEY_BY_OP`, there would be a type error if I tried to access it using `operator`, because `operator` is typed `string` and `MESSAGE_KEY_BY_OP` can't be accessed using any old string, only those four specific strings.

So I create this intermediary `opKey` variable, and type it as `keyof typeof MESSAGE_KEY_BY_OP`. This means that it will become typed as `'not_opened', 'not_sent', 'opened', 'sent'`, i.e. the keys (`keyof`) that exist on the type of (`typeof`) `MESSAGE_KEY_BY_OP`.